### DISCUSSION TITLE
Confused on how to interpret the result and chart

### DISCUSSION THREAD

**Discussion by wakandan** [kernc/backtesting.py#519] [Q&A]
I'm having a very simple strat which buys at every red bar (close<open) to test out stop loss and trades. It doesn't look like the program is working as intended.

```
class MyStrat(TrailingStrategy):
    def init(self):
        super().init()
        self.set_atr_periods(2)
        pass
    
    def next(self):
        super().next()
        close = self.data.Close[-1]
        date = self.data.index[-1]
        adf = self.data.df[-20:].copy()
        current_close = adf.iloc[-1].Close
        adf['down'] = adf.Close<adf.Close.shift(1)
        if adf.down[-1]:# and adf.down[-2]:
            self.buy(sl=current_close*0.98)
            self.position_price = close
            print(f'buy at {self.data.index[-1]}, price = {close}')
            self.last_buy = len(self.data)
            
        
                
bt = Backtest(df[:100], MyStrat, cash=10_000, commission=.0002, margin=0.01, trade_on_close=True)
stats = bt.run()
bt.plot(resample=False)
stats
```

Here're the trades the `print` line printed:
```
buy at 2021-10-05 21:00:00, price = 1.15844
buy at 2021-10-05 21:10:00, price = 1.15861
buy at 2021-10-05 21:15:00, price = 1.15828
buy at 2021-10-05 21:35:00, price = 1.15884
buy at 2021-10-05 21:55:00, price = 1.15923
buy at 2021-10-05 22:15:00, price = 1.16038
buy at 2021-10-05 22:35:00, price = 1.16104
buy at 2021-10-05 22:40:00, price = 1.1609
buy at 2021-10-05 22:45:00, price = 1.16082
buy at 2021-10-05 22:50:00, price = 1.1608
buy at 2021-10-05 23:05:00, price = 1.1609
buy at 2021-10-05 23:10:00, price = 1.16088
buy at 2021-10-05 23:15:00, price = 1.16061
```
These makes sense, because it bought at these red bars. 
However, when printing the `stats._trades`, here is the result

![image](https://user-images.githubusercontent.com/537655/139407559-91481a52-5bfa-492f-b719-570af14a6f81.png)

Only 5 trades are shown, which is very weird. And here is the chart 
![image](https://user-images.githubusercontent.com/537655/139407637-b3179f1b-2a61-4327-8b97-29cb765e2ea5.png)
Note that only 5 trades shown by `stats._trades` are shown.

2 questions:
* Why doesn't any stop loss kick in? 
* Why does it show only 5 trades when in fact in my strat I expected these trades above?

Thanks a lot!

**Comment by wakandan** [comment]
[eurusd_5m.csv](https://github.com/kernc/backtesting.py/files/7445416/eurusd_5m.csv)

Her's the data that I'm using. It's 5m fx data with 5000 bars. Here's how to load it 
```
df = pd.read_csv('eurusd_5m.csv')
df.columns = ['Date', 'Open', 'High', 'Low', 'Close', 'Volume']
df['Date'] = pd.to_datetime(df['Date'], format='%Y/%m/%d, %H:%M:%S')
df = df.set_index('Date')
df
```

**Reply by Buddyli82** [reply]
hmm...you don't seem to be closing your buy positions... you need to have: self.position.close()  somewhere in your code when stop-loss condition is reached.  Where is the stop-loss logic in next()?
